packages/jsii/lib/compiler-watch.ts

```diff
@@ -0,0 +1,155 @@
+    configPath: string,
+    compilerOptions: CompilerOptions,
+    onCompilationComplete: TypeScriptProgramHandler,
+    onStatusChanged?: WatchStatusReporter,
+  ) {
+    const host = createWatchCompilerHost(
+      configPath,
+      {
+        ...projectInfo.tsc,
+        ...compilerOptions,
+        noEmitOnError: false,
+      },
+      { ...sys, getCurrentDirectory() { return projectInfo.projectRoot; } },
+      createSemanticDiagnosticsBuilderProgram,
+      undefined, // reportDiagnostics -> We don't pass this because we have our own reporting mechanism
+      onStatusChanged,
+    );
+
+    if (!host.getDefaultLibLocation) {
+      throw new Error('No default library location was found on the TypeScript compiler host!');
+    }
+    const orig = host.afterProgramCreate?.bind(host);
+    host.afterProgramCreate = async builderProgram => {
+      this.lastResult = await onCompilationComplete(builderProgram.getProgram(), host.getDefaultLibLocation!());
+      if (orig != null) { orig(builderProgram); }
+      this.eventEmitter.emit(CompilerWatch.EVENT_COMPILATION_COMPLETE, this.lastResult, this);
+    };
+
+    this.watch = createWatchProgram(host);
+    this.watchComplete = new Promise((resolve, reject) => {
+      this.onWatchEnd = resolve;
+      this.onWatchFailed = reject;
+    });
+  }
+
+  /**
+   * Adds a listener for the 'result' event that is called with EmitResult.
+   *
+   * @param event    the 'result' event
+   * @param listener the listener to be called with the EmitResult.
+   */
+  public on(event: typeof CompilerWatch.EVENT_COMPILATION_COMPLETE, listener: CompilationCompleteHandler): this {
+    this.eventEmitter.on(event, listener);
+    return this;
+  }
+
+  /**
+   * @param timeoutMillis the maximum amount of time to wait for the watch to complete. When
+   *                      provided, the returned promise will reject if it is not resolved by the
+   *                      end of this amount of time.
+   *
+   * @returns a `Promise` that only resolves once the `.close()` method was invoked.
+   */
+  public async block(timeoutMillis?: number): Promise<EmitResult> {
+    const promises = [this.watchComplete];
+
+    if (timeoutMillis != null) {
+      // This Promise will reject after timeoutMillis
+      promises.push(new Promise<never>((_, reject) => {
+        const id = setTimeout(
+          () => {
+            clearTimeout(id);
+            reject(new Error(`Watch did not terminate within ${timeoutMillis} ms!`));
+          },
+          timeoutMillis,
+        );
+      }));
+    }
+
+    // Wait for one of the `promises` to resolve or reject (whichever comes first)!
+    return Promise.race(promises);
+  }
+
+  /**
+   * Terminates the watch process.
+   */
+  public close(): void {
+    // Calling `this.watch.close()` will "synchronize" the watch, and emit once last time
+    this.eventEmitter.once(CompilerWatch.EVENT_COMPILATION_COMPLETE, this.onWatchEnd);
+    this.watch.close();
```

Comment:
wouldn't it be easier to just expose the raw `watch` object for unit tests to directly access the typescript API instead of duplicating this logic in this layer? Since this is an internal API, I don't see a reason why it can't simply be a leaky API for unit tests, which is a common practice.

Reply:
It's a common practice I dislike, because it encourages poor separation of concerns in my opinion. However, this is not a hill I'm going to die on, so I changed as you're asking.